docs: how does it work (#47)
Open-Attestation/documentation-website: docs/advanced/how-it-works.md

          "location": "c15358f4-f0dc-41c8-abfb-0d030aae3233:string:imaginative-amber-ferret.sandbox.openattestation.com"
        },
        "name": "0de92429-f8d3-47a0-868f-154227a66f40:string:DEMO STORE",
        "tokenRegistry": "89c1f33c-121d-4622-a561-12fb400f2f3f:string:0x8194648f40ED07F841fA357Bf52CBE8D6d7ce48D"
      }
    ]
  },
  "signature": {
    "type": "SHA3MerkleProof",
    "targetHash": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987",
    "proof": [],
    "merkleRoot": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987"
  }
}
```

Multiple interesting things happened that we will dig in:

- A `data` has been created and holds the content of the file provided for wrapping with some weird-looking extra data.
- A `signature` object has been created.

### The `data` object

The first step of wrapping consists of running through and transforming all the object properties provided as input:

1. For each properties
1. Generate a salt [using uuid v4](https://www.npmjs.com/package/uuid) in order to prevent [rainbow table attack](https://en.wikipedia.org/wiki/Rainbow_table).

> 1. Find The type original property.

Reviewer: ```suggestion
1. Find the type original property.
```

Author: done

1. Transform the original value to `<salt>:<original-type>:<original-value>`.

> The shape of the input object remains untouched.

### The `signature` object

#### targetHash

> Once the `data` object has been computed we will be able to create an uniq hash for the document that we will set into `targetHash`:

Reviewer: ```suggestion
Once the `data` object has been computed we will be able to create an unique hash for the document that we will set into `targetHash`:
```

Author: done

        },
        "name": "0de92429-f8d3-47a0-868f-154227a66f40:string:DEMO STORE",
        "tokenRegistry": "89c1f33c-121d-4622-a561-12fb400f2f3f:string:0x8194648f40ED07F841fA357Bf52CBE8D6d7ce48D"
      }
    ]
  },
  "signature": {
    "type": "SHA3MerkleProof",
    "targetHash": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987",
    "proof": [],
    "merkleRoot": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987"
  },
  "privacy": {
    "obfuscatedData": [
      "9d22655fcee6bf3eb10ba280cfa40e662f004a819be0b64e2fe9d0cebba6788f"
    ]
  }
}
```

The `name` field is not available anymore in the `data` object, and the hash associated to it is added into `privacy.obfuscatedData`.

> More importantly, the document remains valid.

The hash added is the one used when computing the [`targetHash`](#targethash). Like said above, we follow the same steps to compute a `targetHash` and verify that a document remained untouched, with one subtle difference: all the hashes available in `privacy.obfuscatedData` are added to the list of computed hashes from the data object before sorting them. That way we can rebuild the `targetHash` with obfuscated values.

> Thanks to data obfuscation a user can decide to selectively decide which data he wants to share.

Reviewer: ```suggestion
Thanks to data obfuscation a user can decide to selectively disclose a subset of data he wants to share.
```

Author: done

### Document Store

> As discussed above the `targetHash` and the `merkleRoot` of a document can be used for issuance. However, the `merkleRoot` is by far the most efficient way.

Reviewer: Don't think we want to give people an idea that they can issue with targetHash, they cant when it's batched. Prefer to speak about 

The issuer can choose to issue documents individually or in a batch...

Author: done

---
id: how-it-works

> title: How it works ?

Reviewer 2: ```suggestion
title: How does it work?
```

> sidebar_label: How it works ?

Reviewer 2: ```suggestion
sidebar_label: How does it work?
```

## Wrapping a document

As a reminder, wrapping a document works on JSON object. A single wrapped document will look like this:

```json
{
  "version": "https://schema.openattestation.com/2.0/schema.json",
  "data": {
    "name": "2f1a9924-bc38-455c-b39e-6420001ad67b:string:Maersk Bill of Lading",
    "issuers": [
      {
        "identityProof": {
          "type": "40caddff-5cd4-477d-adf4-48dcd0a2d761:string:DNS-TXT",
          "location": "c15358f4-f0dc-41c8-abfb-0d030aae3233:string:imaginative-amber-ferret.sandbox.openattestation.com"
        },
        "name": "0de92429-f8d3-47a0-868f-154227a66f40:string:DEMO STORE",
        "tokenRegistry": "89c1f33c-121d-4622-a561-12fb400f2f3f:string:0x8194648f40ED07F841fA357Bf52CBE8D6d7ce48D"
      }
    ]
  },
  "signature": {
    "type": "SHA3MerkleProof",
    "targetHash": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987",
    "proof": [],
    "merkleRoot": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987"
  }
}
```

> Multiple interesting things happened that we will dig in:

Reviewer 2: ```suggestion
A few interesting transformations happened that we will dive into below:
```

> - A `data` has been created and holds the content of the file provided for wrapping with some weird-looking extra data.

Reviewer 2: ```suggestion
- A `data`  key has been created and its value holds the contents of the file previously provided when wrapping, along with some weird-looking extra (hexadecimal) data.
```

- A `signature` object has been created.

### The `data` object

> The first step of wrapping consists of running through and transforming all the object properties provided as input:

Reviewer 2: ```suggestion
The first step of wrapping consists of transforming all the object properties provided as input using the following algorithm:
```

> 1. For each properties

Reviewer 2: ```suggestion
1. For each property
```

1. Generate a salt [using uuid v4](https://www.npmjs.com/package/uuid) in order to prevent [rainbow table attack](https://en.wikipedia.org/wiki/Rainbow_table).

> 1. Find the type original property.

Reviewer 2: ```suggestion
1. Determine the type of original property.
```

Reviewer 2: dunno if you want to mention the types are the same types that are valid json?

1. Transform the original value to `<salt>:<original-type>:<original-value>`.

> The shape of the input object remains untouched.

### The `signature` object

#### targetHash

Once the `data` object has been computed we will be able to create an unique hash for the document that we will set into `targetHash`:

> 1. List each properties' path from the `data` object and associate its value. The path follows the [flatley](https://github.com/antony/flatley) path convention.

Reviewer 2: example?

Author: added

> 1. For each properties' path, compute a hash using the properties' path and value. To compute the hash we use [keccack256](https://en.wikipedia.org/wiki/SHA-3).

Reviewer 2: ```suggestion
1. For each properties' path, compute a hash using the properties' path and value. To compute the hash we use [keccak256](https://en.wikipedia.org/wiki/SHA-3).
```

Reviewer 2: e.g: keccak256("issuers[0].identityProof.type: 40caddff-5cd4-477d-adf4-48dcd0a2d761:string:DNS-TXT")

(i dont remember exactly the flatley format)

Author: We dont do that exactly that's why I didnt provide this (it's wrapped in an empty object)

> 1. Sort all the hashes from the previous step alphabetically and hash them all together: this will provide the `targetHash` of the document. To compute the `targetHash` we also use [keccack256](https://en.wikipedia.org/wiki/SHA-3).

Reviewer 2: ```suggestion
1. Sort all the hashes from the previous step alphabetically and hash them all together: this will provide the `targetHash` of the document. To compute the `targetHash` we also use [keccak256](https://en.wikipedia.org/wiki/SHA-3).
```

> The `targetHash` of a document is an unique identifier.

> We follow exactly the same steps to verify that a document has not been tampered.

Reviewer 2: ```suggestion
Later on, during verification of the document, the same exact steps are performed again to assert that the contents of the document has not been tampered with. This works as the final targetHash will be completely different if any part of the wrapped document is different from the original.
```

> The [document store](/docs/verifiable-document/document-store) is a smart contract on the Ethereum network that records the issuance and revocation status of OpenAttestation documents. It will store hashes of wrapped documents that proves that the owner of the document store really issued the document.

Reviewer 2: ```suggestion
The [document store](/docs/verifiable-document/document-store) is a smart contract on the Ethereum network that records the issuance and revocation status of OpenAttestation documents. It stores the hashes of wrapped documents, which are the records of the owner of the document store having issued the documents.
```

> Imagine that you wrap a thousand of file and had to issue the `targetHash` for each of them. It would be extremely inefficient. That's where the `merkleRoot` will come handy.

Reviewer 2: ```suggestion
Imagine that you wrap a thousand of file and had to issue the `targetHash` for each of them. It would be extremely inefficient. That's where the `merkleRoot` will come in handy.
```

Reviewer 2: seems a bit informal compared to the rest of the doc, but whatever

#### merkleRoot

> Once the `targetHash` of a document is computed, OpenAttestation will determine the `merkleRoot`. The `merkleRoot` value is the merkle root hash computed from the [merkle tree](https://en.wikipedia.org/wiki/Merkle_tree) using the `targetHash` of all the document wrapped together. After computing the merkle tree, the `merkleRoot` associated to a document will be added to it as well as the proofs (intermediate hashes) needed to ensure that the `targetHash` has been used to compute the `merkleRoot`. The proofs are added into the `proof` property.

Reviewer 2: dunno if you wanna talk about what a merkle tree is?
and that each targetHash is a leaf in the tree

and computing the merkle tree using a whole batch of documents (targetHashes)

Author: can add that the targetHash is a leaf of a tree, about merkle tree I would prefer to link to a doc (if there is abything better than wiki)

Author: added `that each targetHash is a leaf in the tree`

> In the document above we can notice that the `targetHash` and the `merkelRoot` are identical and that the `proof` is empty. This is normal and happen when you wrap only one document at a time. Try to wrap at least 2 documents at the same time, and you will see a difference between `targetHash` and the `merkelRoot`, and you will be proofs appended.

Reviewer 2: ```suggestion
In the document above we can notice that the `targetHash` and the `merkelRoot` are identical and that the `proof` is empty. This is normal and happen when you wrap only one document at a time. If you are curious you can experiment with wrapping at least 4 documents in a batch, and you will see a difference between `targetHash` and the `merkelRoot`, and **you will be proofs appended**. (?? i dont understand)
```

Author: you will see proofs appended

> > The `merkleRoot` will always be the same for all the documents wrapped together (at the same time). It will be different for documents wrapped separately.

Reviewer 2: ```suggestion
> The `merkleRoot` will always be the same for all the documents wrapped together (in a batch). It will be different for documents wrapped separately.
```

Now that our batch of documents have a common identifier and that we can prove (thanks to the merkle tree algorithm) that the `targetHash` of a document was used to create a specific `merkleRoot`, we can use the `merkleRoot` in our document store and issue it.

#### Data Obfuscation

> Thanks to the way we compute `targetHash`, OpenAttestation allows for one to obfuscate data they don't want to make public. For this just need to compute the hash of a specific field and add it into the documents. Let's try it with the [CLI](/docs/component/open-attestation-cli) and the document above:

Reviewer 2: thanking merkle tree / targetHash twice sounds a bit weird
maybe "due to the way" ?

Author: changed

> Thanks to the way we compute `targetHash`, OpenAttestation allows for one to obfuscate data they don't want to make public. For this just need to compute the hash of a specific field and add it into the documents. Let's try it with the [CLI](/docs/component/open-attestation-cli) and the document above:

Reviewer 2: ```suggestion
Thanks to the way we compute `targetHash`, OpenAttestation allows for one to obfuscate data they don't want to make public. For this we can simply compute the hash of a specific field and add it into the documents. Let's try it with the [CLI](/docs/component/open-attestation-cli) and the document above:
```

```

The content of `output.json` will be:

```json
{
  "version": "https://schema.openattestation.com/2.0/schema.json",
  "data": {
    "issuers": [
      {
        "identityProof": {
          "type": "40caddff-5cd4-477d-adf4-48dcd0a2d761:string:DNS-TXT",
          "location": "c15358f4-f0dc-41c8-abfb-0d030aae3233:string:imaginative-amber-ferret.sandbox.openattestation.com"
        },
        "name": "0de92429-f8d3-47a0-868f-154227a66f40:string:DEMO STORE",
        "tokenRegistry": "89c1f33c-121d-4622-a561-12fb400f2f3f:string:0x8194648f40ED07F841fA357Bf52CBE8D6d7ce48D"
      }
    ]
  },
  "signature": {
    "type": "SHA3MerkleProof",
    "targetHash": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987",
    "proof": [],
    "merkleRoot": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987"
  },
  "privacy": {
    "obfuscatedData": [
      "9d22655fcee6bf3eb10ba280cfa40e662f004a819be0b64e2fe9d0cebba6788f"
    ]
  }
}
```

The `name` field is not available anymore in the `data` object, and the hash associated to it is added into `privacy.obfuscatedData`.

> More importantly, the document remains valid.

> The hash added is the one used when computing the [`targetHash`](#targethash). Like said above, we follow the same steps to compute a `targetHash` and verify that a document remained untouched, with one subtle difference: all the hashes available in `privacy.obfuscatedData` are added to the list of computed hashes from the data object before sorting them. That way we can rebuild the `targetHash` with obfuscated values.

Reviewer 2: this paragraph is hard to read try to break it down into shorter sentences and simpler words

Author: chaned

> Thanks to data obfuscation a user can decide to selectively disclose a subset of data he wants to share.

Reviewer 2: THANKS AGAIN

Author: changed

### Document Store

> As discussed above, issuance of documents can happen individually or by batch. Choosing to batch documents by far the most efficient way.

Reviewer 2: ```suggestion
As discussed above, issuance of documents can happen individually or by batch. Issuing a batch documents is by far the more efficient way.
```
